internal/proxy/proxy.go

```diff
@@ -39,25 +47,65 @@ func proxyHandler(logger *log.Logger, client *vault.Client) func(http.ResponseWr
 			return
 		}
 
+		cacheKeyHash := ""
+		if shallFetchCache(fwReq, cache) {
+			// Construct the hash for this request to use as the cache key
+			cacheKeyHash, err = makeRequestHash(logger, r, token)
+			if err != nil {
+				logger.Printf("failed to compute request hash: %s", err)
+				http.Error(w, "failed to read request", http.StatusInternalServerError)
+				return
+			}
+			// Check the cache for this request
+			data, err := cache.Get(cacheKeyHash)
```

Comment:
I think we should probably lock an exclusive mutex, and immediately defer the unlock, before we `Get` from the cache. If two requests come in simultaneously, the last one should wait until the first finishes. We could make the mutex specific to the particular key we're retrieving, but that seems likely to be overkill in complexity compared to the typical scale at which a single function will be making Vault requests.

Reply:
Yeah the underlying cache access looks safe. My thinking was more at the application behaviour level. If we don't lock between a cache miss and the underlying request being made, then we could have multiple in-flight Vault requests for the same secret simultaneously. And that could also result in inconsistent results from the cache. e.g. I think currently [this](https://mermaid-js.github.io/mermaid-live-editor/edit#pako:eNp9kU-LwjAQxb9KmLNFdNVDDkLriteFXfaUy5hMtWASTROwFL-7o1S07tLb8H7vvfyZFrQ3BBJqOiVymj4r3AW0yqGOPoj8MRTK5dly-RX8uZFis_4RZjsuvVeu-F--S4x-MR2iFPkgLfr0rmbPXsapppBNph_PaMdWqPf0BzPvhd4b8w7O5ouhxhfMvBcafPbL8WwWZfBW6FstjMBSsFgZ_vNWOSEUxD1ZUiB5NFTeLqpAuQtb09FgpLWpeAMgSzzUNAJM0X83ToOMIdHD1O2tc12uGiidPA) could happen, but we want [this](https://mermaid-js.github.io/mermaid-live-editor/edit/#pako:eNqFkMEKwjAQRH8l7LlF1FsOhVbFq6B4ymVNtrbQJJomoEj_3VQqRax4W-bNg2QeIK0i4NDSNZCRtK7x7FALg9Jbx_L3UQiTp1m2c_Z252y7OTB1mpXWClNMx68ooiOGxnOWf9JXmo5mxKEll84Xy1Ed2AplRV848gkppsWYstJZzWTv_3nnTw8S0OQ01iqO9BCGMQG-Ik0CeDwVlf1PBAjTxWq4KPS0UXWcDHiJTUsJYPB2fzcSuHeB3qVh6KHVPQHhd4Nj) to happen instead. The diagram has two actors, but they could just represent different threads within the same process.

Reply:
We need a way to track inflight requests in order to avoid the situation that Tom described above, where there might be two concurrent inflight requests leading to two of the same requests (e.g. a creds generation request) to be made against the Vault server. 

A global exclusive mutex lock would make the entire caching mechanism synchronous and it'd only be able to process one request at a time. The[ approach that I took for agent](https://github.com/hashicorp/vault/pull/10705/files) to address this issue is to make a second and separate inflight cache that is clears once the first request "wins". The other requests will block (using a channel) until then. This is a way to coordinate concurrent inflight requests of the same type without having to manage locks per request or making the entire cache operate on a global lock.

Reply:
I think allowing multiple inflight requests is more complexity than we need. The scale demands on this proxy are much smaller than for Agent, because each instance of the proxy is dedicated to only one instance of the Lambda function at a time. Given that, I actually don't think a global mutex would be disastrous, but I'm happy to start with a mutex per cache key instead.

Reply:
Appreciate the discussion and the diagrams especially here. I added a mutex per cache key in 6e523bf, and then moved the lock acquisition to the top of the proxyHandler in 8c49e18.

I was thinking about a similar scenario with two requests, but one of them has "recache" set. The "recache" request would skip the cache.Get() (and the locking), but still get the cred from Vault and store it in the cache.